File: implementations/python/lib/eval/positional.py
"""Positional evaluation - bishop pair, rooks on open files, outposts."""

from typing import Tuple
from lib.board import Board
from lib.types import Color, PieceType

BISHOP_PAIR_BONUS: int = 30
ROOK_OPEN_FILE_BONUS: int = 25
ROOK_SEMI_OPEN_FILE_BONUS: int = 15
ROOK_SEVENTH_RANK_BONUS: int = 20
KNIGHT_OUTPOST_BONUS: int = 20


def evaluate(board: Board) -> int:
    """Evaluate positional factors for both sides."""
    score = 0
    
    score += evaluate_color(board, Color.WHITE)
    score -= evaluate_color(board, Color.BLACK)
    
    return score


def evaluate_color(board: Board, color: Color) -> int:
    """Evaluate positional factors for one color."""
    score = 0
    
    if has_bishop_pair(board, color):
        score += BISHOP_PAIR_BONUS
    
    for square in range(64):
        row = square // 8
        col = square % 8
        piece = board.get_piece(row, col)
        
        if piece and piece.color == color:
            if piece.type == PieceType.ROOK:
                score += evaluate_rook(board, square, color)
            elif piece.type == PieceType.KNIGHT:
                score += evaluate_knight(board, square, color)
    
    return score


def has_bishop_pair(board: Board, color: Color) -> bool:
    """Check if player has bishop pair."""
    bishop_count = 0
    
    for square in range(64):
        row = square // 8
        col = square % 8
        piece = board.get_piece(row, col)
        
        if piece and piece.color == color and piece.type == PieceType.BISHOP:
            bishop_count += 1
    
    return bishop_count >= 2


def evaluate_rook(board: Board, square: int, color: Color) -> int:
    """Evaluate rook position."""
    file = square % 8
    rank = square // 8
    bonus = 0
    
    own_pawns, enemy_pawns = count_pawns_on_file(board, file, color)
    
    if own_pawns == 0 and enemy_pawns == 0:
        bonus += ROOK_OPEN_FILE_BONUS
    elif own_pawns == 0:
        bonus += ROOK_SEMI_OPEN_FILE_BONUS
    
    seventh_rank = 6 if color == Color.WHITE else 1
    if rank == seventh_rank:
        bonus += ROOK_SEVENTH_RANK_BONUS
    
    return bonus


def evaluate_knight(board: Board, square: int, color: Color) -> int:
    """Evaluate knight position."""
    if is_outpost(board, square, color):
        return KNIGHT_OUTPOST_BONUS
    return 0


def is_outpost(board: Board, square: int, color: Color) -> bool:
    """Check if knight is on an outpost."""
    file = square % 8
    rank = square // 8
    
    if not is_protected_by_pawn(board, square, color):
        return False
    
    if can_be_attacked_by_enemy_pawn(board, square, file, rank, color):
        return False
    
    return True


def is_protected_by_pawn(board: Board, square: int, color: Color) -> bool:
    """Check if square is protected by own pawn."""
    file = square % 8
    rank = square // 8
    
    behind_rank = rank - 1 if color == Color.WHITE else rank + 1
    
    if not (0 <= behind_rank < 8):
        return False
    
    for adjacent_file in [file - 1, file + 1]:
        if 0 <= adjacent_file < 8:
            piece = board.get_piece(behind_rank, adjacent_file)
            if piece and piece.color == color and piece.type == PieceType.PAWN:
                return True
    
    return False


def can_be_attacked_by_enemy_pawn(board: Board, square: int, file: int, rank: int, color: Color) -> bool:
    """Check if square can be attacked by enemy pawn."""
    if color == Color.WHITE:
        ahead_ranks = range(rank + 1, 8)
    else:
        ahead_ranks = range(0, rank)
    
    for check_rank in ahead_ranks:
        for adjacent_file in [file - 1, file + 1]:
            if 0 <= adjacent_file < 8:
                piece = board.get_piece(check_rank, adjacent_file)
                if piece and piece.color != color and piece.type == PieceType.PAWN:
                    return True
    
    return False


def count_pawns_on_file(board: Board, file: int, color: Color) -> Tuple[int, int]:
    """Count own and enemy pawns on a file."""
    own_pawns = 0
    enemy_pawns = 0
    
    for rank in range(8):
        piece = board.get_piece(rank, file)
        if piece and piece.type == PieceType.PAWN:
            if piece.color == color:
                own_pawns += 1
            else:
                enemy_pawns += 1
    
    return own_pawns, enemy_pawns

File: implementations/dart/lib/eval/positional.dart
import 'package:chess_engine/chess_engine.dart';

const int bishopPairBonus = 30;
const int rookOpenFileBonus = 25;
const int rookSemiOpenFileBonus = 15;
const int rookSeventhRankBonus = 20;
const int knightOutpostBonus = 20;

int evaluate(Board board) {
  int score = 0;
  
  score += _evaluate_color(board, PieceColor.white);
  score -= _evaluate_color(board, PieceColor.black);
  
  return score;
}

int _evaluate_color(Board board, PieceColor color) {
  int score = 0;
  
  if (_has_bishop_pair(board, color)) {
    score += bishopPairBonus;
  }
  
  for (int square = 0; square < 64; square++) {
    final row = square ~/ 8;
    final col = square % 8;
    final piece = board.squares[row][col];
    
    if (piece != null && piece.color == color) {
      switch (piece.type) {
        case PieceType.rook:
          score += _evaluate_rook(board, square, color);
          break;
        case PieceType.knight:
          score += _evaluate_knight(board, square, color);
          break;
        default:
          break;
      }
    }
  }
  
  return score;
}

bool _has_bishop_pair(Board board, PieceColor color) {
  int bishopCount = 0;
  
  for (int square = 0; square < 64; square++) {
    final row = square ~/ 8;
    final col = square % 8;
    final piece = board.squares[row][col];
    
    if (piece != null && piece.color == color && piece.type == PieceType.bishop) {
      bishopCount++;
    }
  }
  
  return bishopCount >= 2;
}

int _evaluate_rook(Board board, int square, PieceColor color) {
  final file = square % 8;
  final rank = square ~/ 8;
  int bonus = 0;
  
  final pawns = _count_pawns_on_file(board, file, color);
  final ownPawns = pawns.$1;
  final enemyPawns = pawns.$2;
  
  if (ownPawns == 0 && enemyPawns == 0) {
    bonus += rookOpenFileBonus;
  } else if (ownPawns == 0) {
    bonus += rookSemiOpenFileBonus;
  }
  
  final seventhRank = color == PieceColor.white ? 6 : 1;
  if (rank == seventhRank) {
    bonus += rookSeventhRankBonus;
  }
  
  return bonus;
}

int _evaluate_knight(Board board, int square, PieceColor color) {
  if (_is_outpost(board, square, color)) {
    return knightOutpostBonus;
  } else {
    return 0;
  }
}

bool _is_outpost(Board board, int square, PieceColor color) {
  final file = square % 8;
  final rank = square ~/ 8;
  
  final protectedByPawn = _is_protected_by_pawn(board, square, color);
  if (!protectedByPawn) {
    return false;
  }
  
  final cannotBeAttacked = !_can_be_attacked_by_enemy_pawn(board, square, file, rank, color);
  
  return protectedByPawn && cannotBeAttacked;
}

bool _is_protected_by_pawn(Board board, int square, PieceColor color) {
  final file = square % 8;
  final rank = square ~/ 8;
  
  final behindRank = color == PieceColor.white
      ? (rank - 1).clamp(0, 7)
      : (rank + 1).clamp(0, 7);
  
  for (final adjacentFile in [(file - 1).clamp(0, 7), (file + 1).clamp(0, 7)]) {
    if (adjacentFile != file) {
      final checkSquare = behindRank * 8 + adjacentFile;
      final row = checkSquare ~/ 8;
      final col = checkSquare % 8;
      final piece = board.squares[row][col];
      
      if (piece != null && piece.color == color && piece.type == PieceType.pawn) {
        return true;
      }
    }
  }
  
  return false;
}

bool _can_be_attacked_by_enemy_pawn(Board board, int square, int file, int rank, PieceColor color) {
  final aheadRanks = color == PieceColor.white
      ? List<int>.generate(8 - rank - 1, (i) => rank + 1 + i)
      : List<int>.generate(rank, (i) => i);
  
  for (final checkRank in aheadRanks) {
    for (final adjacentFile in [(file - 1).clamp(0, 7), (file + 1).clamp(0, 7)]) {
      if (adjacentFile != file) {
        final checkSquare = checkRank * 8 + adjacentFile;
        final row = checkSquare ~/ 8;
        final col = checkSquare % 8;
        final piece = board.squares[row][col];
        
        if (piece != null && piece.color != color && piece.type == PieceType.pawn) {
          return true;
        }
      }
    }
  }
  
  return false;
}

(int, int) _count_pawns_on_file(Board board, int file, PieceColor color) {
  int ownPawns = 0;
  int enemyPawns = 0;
  
  for (int rank = 0; rank < 8; rank++) {
    final square = rank * 8 + file;
    final row = square ~/ 8;
    final col = square % 8;
    final piece = board.squares[row][col];
    
    if (piece != null && piece.type == PieceType.pawn) {
      if (piece.color == color) {
        ownPawns++;
      } else {
        enemyPawns++;
      }
    }
  }
  
  return (ownPawns, enemyPawns);
}
